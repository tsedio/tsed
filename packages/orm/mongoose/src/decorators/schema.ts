import {decoratorTypeOf, StoreMerge, useDecorators} from "@tsed/core";
import {registerProvider} from "@tsed/di";
import {Property} from "@tsed/schema";
import {SchemaTypeOptions} from "mongoose";
import {MONGOOSE_SCHEMA} from "../constants";
import {MongooseSchemaOptions} from "../interfaces";
import {getSchema, getSchemaToken} from "../utils/createSchema";

/**
 * Define a class as a Mongoose Schema ready to be used to compose other schemes and models.
 *
 * ### Example
 *
 * ```typescript
 * @MongooseSchema()
 * export class EventSchema {
 *   @Property()
 *   field: string;
 * }
 * ```
 *
 * ### Options
 *
 * - `schemaOptions` (mongoose.SchemaOptions): Option to configure the schema behavior.
 *
 * @param {MongooseSchemaOptions | undefined} options
 * @returns {(target: any) => void}
 * @decorator
 * @mongoose
 * @property
 * @class
 */
export function Schema(options?: MongooseSchemaOptions): (target: any) => void;
export function Schema(definition: SchemaTypeOptions<any>): Function;
export function Schema(options: MongooseSchemaOptions | SchemaTypeOptions<any> = {}) {
  return (...parameters: any[]) => {
    switch (decoratorTypeOf(parameters)) {
      case "property":
        return useDecorators(Property(), StoreMerge(MONGOOSE_SCHEMA, options))(parameters[0], parameters[1], parameters[2]);

      case "class":
        const {token} = getSchemaToken(parameters[0], options);

        registerProvider({
          provide: token,
          deps: [],
          useFactory() {
            return getSchema(parameters[0], options as any);
          }
        });
        break;
    }
  };
}

/**
 * Define a class as a Mongoose Schema ready to be used to compose other schemes and models.
 *
 * ### Example
 *
 * ```typescript
 * @MongooseSchema()
 * export class EventSchema {
 *   @Property()
 *   field: string;
 * }
 * ```
 *
 * ### Options
 *
 * - `schemaOptions` (mongoose.SchemaOptions): Option to configure the schema behavior.
 *
 * @param {MongooseSchemaOptions | undefined} options
 * @returns {(target: any) => void}
 * @decorator
 * @mongoose
 * @property
 * @class
 */
export function MongooseSchema(options?: MongooseSchemaOptions): (target: any) => void;
export function MongooseSchema(definition: SchemaTypeOptions<any>): Function;
export function MongooseSchema(options: MongooseSchemaOptions | SchemaTypeOptions<any> = {}) {
  return Schema(options as any);
}
