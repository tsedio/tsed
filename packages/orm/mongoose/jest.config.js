// For a detailed explanation regarding each configuration property, visit:
// https://jestjs.io/docs/en/configuration.html

module.exports = {
  ...require("@tsed/jest-config")(__dirname, "mongoose"),
  coverageThreshold: {
    global: {
      statements: 99.5,
      branches: 95.48,
      functions: 100,
      lines: 99.47
    }
  }
};
