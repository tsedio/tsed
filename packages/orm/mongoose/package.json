{
  "name": "@tsed/mongoose",
  "description": "Mongoose package for Ts.ED framework",
  "type": "commonjs",
  "version": "7.79.3",
  "keywords": [
    "tsed",
    "mongoose",
    "typescript",
    "decorators",
    "orm"
  ],
  "source": "./src/index.ts",
  "main": "./lib/cjs/index.js",
  "module": "./lib/esm/index.js",
  "typings": "./lib/types/index.d.ts",
  "exports": {
    ".": {
      "types": "./lib/types/index.d.ts",
      "import": "./lib/esm/index.js",
      "default": "./lib/cjs/index.js"
    }
  },
  "scripts": {
    "build": "yarn barrels && yarn build:ts",
    "barrels": "barrels",
    "test": "vitest run",
    "build:ts": "tsc --build tsconfig.json",
    "test:ci": "vitest run --coverage.thresholds.autoUpdate=true"
  },
  "dependencies": {
    "change-case": "4.1.2",
    "tslib": "2.6.1"
  },
  "devDependencies": {
    "@tsed/ajv": "workspace:*",
    "@tsed/barrels": "workspace:*",
    "@tsed/core": "workspace:*",
    "@tsed/di": "workspace:*",
    "@tsed/json-mapper": "workspace:*",
    "@tsed/logger": ">=6.7.5",
    "@tsed/schema": "workspace:*",
    "@tsed/typescript": "workspace:*",
    "eslint": "^8.57.0",
    "mongoose": "6.12.7",
    "typescript": "4.9.5",
    "vitest": "2.0.4"
  },
  "peerDependencies": {
    "@tsed/ajv": "7.79.3",
    "@tsed/core": "7.79.3",
    "@tsed/di": "7.79.3",
    "@tsed/json-mapper": "7.79.3",
    "@tsed/logger": ">=6.7.5",
    "@tsed/schema": "7.79.3",
    "mongoose": ">=6.0.0"
  }
}
