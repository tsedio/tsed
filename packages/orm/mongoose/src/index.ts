export * from "./interfaces";
export * from "./constants";
export * from "./services/MongooseService";
export * from "./MongooseModule";
export * from "./utils";
export * from "./decorators";
