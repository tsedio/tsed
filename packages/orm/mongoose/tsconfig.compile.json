{
  "extends": "../../tsconfig.compile.json",
  "compilerOptions": {
    "rootDir": "src",
    "outDir": "lib"
  },
  "exclude": [
    "node_modules",
    "test",
    "lib",
    "**/*.spec.ts"
  ]
}
