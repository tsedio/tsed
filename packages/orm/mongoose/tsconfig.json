{
  "extends": "@tsed/typescript/tsconfig.node.json",
  "compilerOptions": {
    "baseUrl": ".",
    "noEmit": true
  },
  "include": [],
  "references": [
    {
      "path": "../../specs/ajv/tsconfig.json"
    },
    {
      "path": "../../core/tsconfig.json"
    },
    {
      "path": "../../di/tsconfig.json"
    },
    {
      "path": "../../specs/json-mapper/tsconfig.json"
    },
    {
      "path": "../../specs/schema/tsconfig.json"
    },
    {
      "path": "../testcontainers-mongo/tsconfig.json"
    },
    {
      "path": "./tsconfig.cjs.json"
    },
    {
      "path": "./tsconfig.esm.json"
    },
    {
      "path": "./tsconfig.spec.json"
    }
  ]
}
