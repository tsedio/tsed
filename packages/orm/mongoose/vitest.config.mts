// @ts-ignore
import {presets} from "@tsed/vitest/presets";
import {defineConfig} from "vitest/config";

export default defineConfig(
  {
    ...presets,
    test: {
      ...presets.test,
      // @ts-ignore
      globalSetup: [import.meta.resolve("@tsed/testcontainers-mongo/vitest/setup")],
      coverage: {
        ...presets.test.coverage,
        thresholds: {
          statements: 96.29,
          branches: 93.72,
          functions: 100,
          lines: 96.29
        }
      }
    }
  }
);
