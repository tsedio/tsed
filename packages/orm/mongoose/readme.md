<p style="text-align: center" align="center">
 <a href="https://tsed.io" target="_blank"><img src="https://tsed.io/tsed-og.png" width="200" alt="Ts.ED logo"/></a>
</p>

<div align="center">
   <h1>Mongoose</h1>

[![Build & Release](https://github.com/tsedio/tsed/workflows/Build%20&%20Release/badge.svg)](https://github.com/tsedio/tsed/actions?query=workflow%3A%22Build+%26+Release%22)
[![PR Welcome](https://img.shields.io/badge/PRs-welcome-brightgreen.svg)](https://github.com/tsedio/tsed/blob/master/CONTRIBUTING.md)
[![Coverage Status](https://coveralls.io/repos/github/tsedio/tsed/badge.svg?branch=production)](https://coveralls.io/github/tsedio/tsed?branch=production)
[![npm version](https://badge.fury.io/js/%40tsed%2Fcommon.svg)](https://badge.fury.io/js/%40tsed%2Fcommon)
[![semantic-release](https://img.shields.io/badge/%20%20%F0%9F%93%A6%F0%9F%9A%80-semantic--release-e10079.svg)](https://github.com/semantic-release/semantic-release)
[![code style: prettier](https://img.shields.io/badge/code_style-prettier-ff69b4.svg?style=flat-square)](https://github.com/prettier/prettier)
[![github](https://img.shields.io/static/v1?label=Github%20sponsor&message=%E2%9D%A4&logo=GitHub&color=%23fe8e86)](https://github.com/sponsors/romakita)
[![opencollective](https://img.shields.io/static/v1?label=OpenCollective%20sponsor&message=%E2%9D%A4&logo=OpenCollective&color=%23fe8e86)](https://opencollective.com/tsed)

</div>

<div align="center">
  <a href="https://tsed.io/">Website</a>
  <span>&nbsp;&nbsp;‚Ä¢&nbsp;&nbsp;</span>
  <a href="https://tsed.io/getting-started/">Getting started</a>
  <span>&nbsp;&nbsp;‚Ä¢&nbsp;&nbsp;</span>
  <a href="https://api.tsed.io/rest/slack/tsedio/tsed">Slack</a>
  <span>&nbsp;&nbsp;‚Ä¢&nbsp;&nbsp;</span>
  <a href="https://twitter.com/TsED_io">Twitter</a>
</div>

<hr />

A package of Ts.ED framework. See website: https://tsed.io/tutorials/mongoose.html

## Feature

Currently, `@tsed/mongoose` allows you:

- Configure one or more MongoDB database connections via the `@ServerSettings` configuration.
  All databases will be initialized when the server starts during the server's `OnInit` phase.
- Declare a Model from a class with annotation.
- Declare inhertitated models in a single collection via `@DiscriminatorKey`.
- Add a plugin, PreHook method and PostHook on your model.
- Inject a Model to a Service, Controller, Middleware, etc...

> Note: `@tsed/mongoose` use the JsonSchema and his decorators to generate the mongoose schema.

## Installation

Before using the `@tsed/mongoose` package, we need to install the [mongoose](https://www.npmjs.com/package/mongoose) module.

```bash
npm install --save mongoose
npm install --save @tsed/mongoose
npm install --save-dev @types/mongoose
```

Then import `@tsed/mongoose` in your Server:

```typescript
import {Configuration} from "@tsed/common";
import "@tsed/mongoose"; // import mongoose ts.ed module

@Configuration({
  mongoose: [
    {
      id: "default",
      url: "mongodb://127.0.0.1:27017/db1",
      connectionOptions: {}
    }
  ]
})
export class Server {}
```

## Multi databases

The mongoose module of Ts.ED Mongoose allows to configure several basic connections to MongoDB.
Here is an example configuration:

```typescript
import {Configuration} from "@tsed/common";
import "@tsed/mongoose"; // import mongoose ts.ed module

@Configuration({
  mongoose: [
    {
      id: "default",
      url: "mongodb://127.0.0.1:27017/db1",
      connectionOptions: {}
    },
    {
      id: "default",
      url: "mongodb://127.0.0.1:27017/db2",
      connectionOptions: {}
    }
  ]
})
export class Server {}
```

## MongooseService

MongooseService let you to retrieve an instance of Mongoose.Connection.

```typescript
import {Service} from "@tsed/common";
import {MongooseService} from "@tsed/mongoose";

@Service()
export class MyService {
  constructor(mongooseService: MongooseService) {
    mongooseService.get(); // return the default instance of Mongoose.
    // If you have one or more database configured with Ts.ED
    mongooseService.get("default");
    mongooseService.get("db2");
  }
}
```

## Declaring a Model

By default, Ts.ED mongoose will reuse the metadata stored by the decorators dedicated
to describe a JsonSchema. This decorators come from the `@tsed/common` package.

Here a model example:

```typescript
import {Minimum, Maximum, MaxLength, MinLength, Enum, Pattern, Required, CollectionOf} from "@tsed/common";
import {Model, Unique, Indexed, Ref, ObjectID} from "@tsed/mongoose";

enum Categories {
  CAT1 = "cat1",
  CAT2 = "cat2"
}

@Model({dbName: "default"}) // dbName is optional. By default dbName is equal to default
export class MyModel {
  @ObjectID()
  _id: string;

  @Unique()
  @Required()
  unique: string;

  @Indexed()
  @MinLength(3)
  @MaxLength(50)
  indexed: string;

  @Minimum(0)
  @Maximum(100)
  rate: Number;

  @Enum(Categories)
  // or @Enum("type1", "type2")
  category: Categories;

  @Pattern(/[a-z]/) // equivalent of match field in mongoose
  pattern: String;

  @CollectionOf(String)
  arrayOf: string[];

  @Ref(OtherModel)
  ref: Ref<OtherModel>;

  @Ref(OtherModel)
  refs: Ref<OtherModel>[];
}
```

## Inject model

```typescript
import {Service, Inject} from "@tsed/common";
import {MongooseModel} from "@tsed/mongoose";
import {MyModel} from "./models/MyModel";

@Service()
export class MyService {
  constructor(@Inject(MyModel) private model: MongooseModel<MyModel>): MyModel {
    console.log(model); // Mongoose.model class
  }

  async save(obj: MyModel): MongooseModel<MyModel> {
    const doc = new this.model(obj);
    await doc.save();

    return doc;
  }

  async find(query: any) {
    const list = await this.model.find(query).exec();

    console.log(list);

    return list;
  }
}
```

## Register hook

Mongoose allows the developer to add pre and post [hooks / middlewares](http://mongoosejs.com/docs/middleware.html) to the schema.
With this it is possible to add document transformations and observations before or after validation, save and more.

Ts.ED provide class decorator to register middlewares on the pre and post hook.

### Pre hook

We can simply attach a `@PreHook` decorator to your model class and
define the hook function like you normally would in Mongoose.

```typescript
import {Required} from "@tsed/common";
import {PreHook, Model, ObjectID} from "@tsed/mongoose";

@Model()
@PreHook("save", (car: CarModel, next) => {
  if (car.model === "Tesla") {
    car.isFast = true;
  }
  next();
})
export class CarModel {
  @ObjectID()
  _id: string;

  @Required()
  model: string;

  @Required()
  isFast: boolean;

  // or Prehook on static method
  @PreHook("save")
  static preSave(car: CarModel, next) {
    if (car.model === "Tesla") {
      car.isFast = true;
    }
    next();
  }
}
```

This will execute the pre-save hook each time a `CarModel` document is saved.

### Post hook

We can simply attach a `@PostHook` decorator to your model class and
define the hook function like you normally would in Mongoose.

```typescript
import {ObjectID, Required} from "@tsed/common";
import {PostHook, Model} from "@tsed/mongoose";

@Model()
@PostHook("save", (car: CarModel) => {
  if (car.topSpeedInKmH > 300) {
    console.log(car.model, "is fast!");
  }
})
export class CarModel {
  @ObjectID()
  _id: string;

  @Required()
  model: string;

  @Required()
  isFast: boolean;

  // or Prehook on static method
  @PostHook("save")
  static postSave(car: CarModel) {
    if (car.topSpeedInKmH > 300) {
      console.log(car.model, "is fast!");
    }
  }
}
```

This will execute the post-save hook each time a `CarModel` document is saved.

## Plugin

Using the `plugin` decorator enables the developer to attach various Mongoose plugins to the schema.
Just like the regular `schema.plugin()` call, the decorator accepts 1 or 2 parameters: the plugin itself, and an optional configuration object.
Multiple `plugin` decorator can be used for a single model class.

```typescript
import {Service} from "@tsed/common";
import {MongoosePlugin, Model, MongooseModel} from "@tsed/mongoose";
import * as findOrCreate from 'mongoose-findorcreate';

@Model()
@MongoosePlugin(findOrCreate)
class UserModel {
  // this isn't the complete method signature, just an example
  static findOrCreate(condition: InstanceType<User>):
    Promise<{ doc: InstanceType<User>, created: boolean }>;
}

@Service()
class UserService {
    constructor(@Inject(UserModel) userModel: MongooseModel<UserModel>) {
        userModel.findOrCreate({ ... }).then(findOrCreateResult => {
          ...
        });
    }
}
```

## Discriminators

Set the `@DiscriminatorKey` decorator on a property in the parent class to define the name of the field for the discriminator value.

Extend the child model classes from the parent class. By default the value for the discriminator field is the class name but it can be overwritten via the `discriminatorValue` option on the model.

```typescript
@Model()
class EventModel {
  @ObjectID()
  _id: string;

  @Required()
  time: Date = new Date();

  @DiscriminatorKey()
  type: string;
}

@Model()
class ClickedLinkEventModel extends EventModel {
  @Required()
  url: string;
}

@Model({discriminatorValue: "signUpEvent"})
class SignedUpEventModel extends EventModel {
  @Required()
  user: string;
}
```

For further information, please refer to the [mongoose documentation about discriminators](https://mongoosejs.com/docs/discriminators.html).

## Contributors

Please read [contributing guidelines here](https://tsed.io/CONTRIBUTING.html)

<a href="https://github.com/tsedio/ts-express-decorators/graphs/contributors"><img src="https://opencollective.com/tsed/contributors.svg?width=890" /></a>

## Backers

Thank you to all our backers! üôè [[Become a backer](https://opencollective.com/tsed#backer)]

<a href="https://opencollective.com/tsed#backers" target="_blank"><img src="https://opencollective.com/tsed/backers.svg?width=890"></a>

## Sponsors

Support this project by becoming a sponsor. Your logo will show up here with a link to your website. [[Become a sponsor](https://opencollective.com/tsed#sponsor)]

## License

The MIT License (MIT)

Copyright (c) 2016 - 2022 Romain Lenzotti

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the "Software"), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
