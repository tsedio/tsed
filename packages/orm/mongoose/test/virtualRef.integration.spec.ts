import {Controller, Get, Inject, PlatformTest} from "@tsed/common";
import {getValue} from "@tsed/core";
import {serialize} from "@tsed/json-mapper";
import {Model, MongooseModel, ObjectID, VirtualRef, VirtualRefs} from "@tsed/mongoose";
import {PlatformExpress} from "@tsed/platform-express";
import {CollectionOf, getJsonSchema, Groups, Integer, Required} from "@tsed/schema";
import {TestMongooseContext} from "@tsed/testing-mongoose";
import SuperTest from "supertest";
import {Server} from "./helpers/Server";

@Model()
export class GalaxiesModel {
  @ObjectID("id")
  @Groups("!internal")
  _id?: string;

  @Required()
  name: string;

  @Groups("!internal")
  spaceId?: string;
}

@Model()
export class SpacesModel {
  @ObjectID("id")
  _id?: string;

  @Required()
  name: string;

  @VirtualRef({
    ref: GalaxiesModel,
    localField: "_id",
    foreignField: "spaceId",
    justOne: false
  })
  @CollectionOf(GalaxiesModel)
  galaxies?: VirtualRefs<GalaxiesModel>;

  @VirtualRef({
    ref: GalaxiesModel,
    localField: "_id",
    foreignField: "spaceId",
    count: true
  })
  @Groups("!internal")
  @Integer()
  galaxyCount?: number;
}

@Controller("/spaces")
class SpacesController {
  @Inject(SpacesModel)
  spaces: MongooseModel<SpacesModel>;

  @Inject(GalaxiesModel)
  galaxies: MongooseModel<GalaxiesModel>;

  @Get("/")
  getSpaces() {
    return this.spaces.find({}, {}, {populate: ["galaxies", "galaxyCount"]});
  }
}

describe("Mongoose", () => {
  describe("Ref", () => {
    let request: SuperTest.SuperTest<SuperTest.Test>;

    beforeEach(
      TestMongooseContext.bootstrap(Server, {
        platform: PlatformExpress,
        mount: {
          "/rest": [SpacesController]
        }
      })
    );
    beforeEach(async () => {
      const controller = PlatformTest.get<SpacesController>(SpacesController)!;

      const space1 = new controller.spaces({name: "Space1"});
      await space1.save();

      const galaxy1 = new controller.galaxies({name: "galaxy1", spaceId: space1._id});
      await galaxy1.save();
    });
    beforeEach(() => {
      request = SuperTest(PlatformTest.callback());
    });

    afterEach(TestMongooseContext.reset);

    it("should return the json-schema of the given model", () => {
      expect(getJsonSchema(SpacesModel)).toMatchSnapshot();
    });

    it("should map the document to POO", async () => {
      const controller = PlatformTest.get<SpacesController>(SpacesController)!;
      const result = await controller.getSpaces();

      expect(getValue(result, "0.galaxies.0.name")).toEqual("galaxy1");
      expect(serialize(result, {virtuals: ["galaxies", "galaxyCount"]})).toMatchObject([
        {
          name: "Space1",
          galaxyCount: 1,
          galaxies: [
            {
              name: "galaxy1"
            }
          ]
        }
      ]);
    });

    it("GET /spaces", async () => {
      const {body: list} = await request.get(`/rest/spaces`);

      expect(list).toMatchObject([
        {
          name: "Space1",
          galaxyCount: 1,
          galaxies: [
            {
              name: "galaxy1"
            }
          ]
        }
      ]);
    });
  });
});
