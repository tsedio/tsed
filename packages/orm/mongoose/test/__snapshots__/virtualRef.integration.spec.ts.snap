// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Mongoose > VirtualRef > should return the json-schema of the given model 1`] = `
{
  "definitions": {
    "GalaxiesModel": {
      "properties": {
        "id": {
          "description": "An ObjectID",
          "examples": [
            "5ce7ad3028890bd71749d477",
          ],
          "pattern": "^[0-9a-fA-F]{24}$",
          "type": "string",
        },
        "name": {
          "minLength": 1,
          "type": "string",
        },
        "spaceId": {
          "type": "string",
        },
      },
      "required": [
        "name",
      ],
      "type": "object",
    },
  },
  "properties": {
    "galaxies": {
      "items": {
        "$ref": "#/definitions/GalaxiesModel",
      },
      "type": "array",
    },
    "galaxyCount": {
      "multipleOf": 1,
      "type": "number",
    },
    "id": {
      "description": "An ObjectID",
      "examples": [
        "5ce7ad3028890bd71749d477",
      ],
      "pattern": "^[0-9a-fA-F]{24}$",
      "type": "string",
    },
    "name": {
      "minLength": 1,
      "type": "string",
    },
  },
  "required": [
    "name",
  ],
  "type": "object",
}
`;
